kani-compiler/src/codegen_cprover_gotoc/codegen/place.rs

```diff
@@ -642,8 +642,33 @@ impl<'tcx> GotocCtx<'tcx> {
         let projection =
             unwrap_or_return_codegen_unimplemented!(self, self.codegen_place_stable(place));
         if self.use_thin_pointer_stable(place_ty) {
-            // Just return the address of the place dereferenced.
-            projection.goto_expr.address_of()
+            // For non-parameter objects with ZST rustc does not necessarily generate
+            // individual objects.
```

Comment:
1. In contrast to C, provenance does not seem all that important in Rust. https://doc.rust-lang.org/std/ptr/fn.eq.html does not rule out comparing pointers of different provenance. https://github.com/rust-lang/unsafe-code-guidelines/issues/239 asks that question, but it's still left open.
2. I am not sure whether any guarantees about pointer identity are provided. https://github.com/rust-lang/unsafe-code-guidelines/issues/406 has a related discussion.
3. Is there any documented specification that a static and a local variable _cannot_ have the same address?

Reply:
1. Rust memory model is still being developed, and there are different proposals for how to handle provenance. There's a few proposals right now, such as Strict Provenance: https://doc.rust-lang.org/std/ptr/index.html#strict-provenance. A ZST pointer to a deallocated memory is also invalid, that's because of provenance: "Even for operations of size zero, the pointer must not be pointing to deallocated memory, i.e., deallocation makes pointers invalid even for zero-sized operations."
2. That's just related to constants, not to ZST.
3. 
> All references to the static refer to the same memory location. https://doc.rust-lang.org/reference/items/static-items.html
> A local variable (or stack-local allocation) holds a value directly, allocated within the stack's memory. The value is a part of the stack frame. https://doc.rust-lang.org/reference/variables.html

Reply:
Also resolving this one as a) parameters no longer get special treatment and b) the discussion in https://github.com/rust-lang/unsafe-code-guidelines/issues/503 suggests that this over-approximation is in line with the current specification.